WIP: Feat/push notifications

Sol3iman opened the pull request:
- The way the rate oracle information is grabbed for handleIrsInstanceDeployed overwrites the previous instance 
- This part of the code should be refactored to load the rateOracle instance and not overwrite it with a new one which has identical entries
- These comments are just to keep track of the WIP
- schema has not been generated, yet.

I have made changes and tried to deploy on a test subgaph but get these errors. @0xZenus do you know why I can't access object properties from NOTIFICATION_CONFIGS?

Edit:
The above errors have been fixed

Edit:
This subgraph is deployed separately under the name voltz-push-notification-test subgraph. Merge this PR when the initiative is re-opened in Jan 2023

Sol3iman committed:
feat: when a position is liquidated and new pool is launched an alert is sent to push channel
	- handleLiquidatePosition contains sendEPNSNotification() method to send notif when a position is liquidated WIP
	- handleIrsInstanceDeployed contains sendEPNSNotification() method to send notif when new pool is launched
	- mapping file is temporary
	- EPNSNotification file is needed to handle the notification sending process
	- added getProtocolPrefix to retrieve the market of pools

Sol3iman committed:
feat: liquidation alert specifies the pool underlying token and yield-bearing token
	- new pool alert specifies which token and maturity dates
	- mappings commented code removed
	- still WIP

Sol3iman committed:
fix: remove mapping file and export subgraphID from handleLiquidatePosition, refactor file structure

avniculae commented on src/mappings/notifications/EPNSNotification.ts:
```diff
@@ -0,0 +1,31 @@
+
+export function sendEPNSNotification(recipient: string, notification: string): void {
+  const id1 = subgraphID;
+  log.info('New id of EpnsNotificationCounter is: {}', [id1]);
```
This is not a new ID, right? EpnsNotificationCounter would always have the subgraphID ID. I think you can just remove the id1 variable, as the name is not suggestive and we can just use subgraphID

avniculae commented on src/mappings/notifications/EPNSNotification.ts:
```diff
@@ -0,0 +1,31 @@
+  epnsNotificationCounter.totalCount = epnsNotificationCounter.totalCount.plus(BigInt.fromI32(1));
+
+  const count = epnsNotificationCounter.totalCount.toHexString();
+  const id2 = `${subgraphID}+${count}`;
```
use a more suggestive name

avniculae commented on src/mappings/notifications/EPNSNotification.ts:
```diff
@@ -0,0 +1,31 @@
+  log.info('New id of EpnsPushNotification is: {}', [id2]);
+
+  let epnsPushNotification = EpnsPushNotification.load(id2);
+  if (epnsPushNotification == null) {
```
this should always be null, right? shall we return if it's not?

Sol3iman commented on src/mappings/marginEngine/handleLiquidatePosition.ts:
```diff
@@ -46,6 +49,45 @@ function handleLiquidatePosition(event: PositionLiquidation): void {
+
+  sendEPNSNotification(recipient, notification);
+}
+export const subgraphID = 'voltzprotocol/voltz-goerli'; // change to mainnet when deploying to mainnet subgraph, maybe keep this in the main mappings file?
```
I agree, will do this

Sol3iman replied:
done

Sol3iman commented on src/mappings/factory/handleIrsInstanceDeployed.ts:
```diff
@@ -52,6 +55,36 @@ function handleIrsInstanceDeployed(event: IrsInstance): void {
   amm.tickSpacing = BigInt.fromI32(event.params.tickSpacing);
   amm.save();
+
+  const recipient = '0x45556408e543158f74403e882E3C8c23eCD9f73'; // goerli test channel address
```
agreed

Sol3iman replied:
done

Sol3iman commented on src/mappings/marginEngine/handleLiquidatePosition.ts:
```diff
@@ -46,6 +49,45 @@ function handleLiquidatePosition(event: PositionLiquidation): void {
+  const secret = 'null';
+  const cta = 'https://app.voltz.xyz/';
+
+  const notification = `{
```
agreed

0xZenus replied:
Either that, or I would move this to a function in the same file (i.e. pushLiquidationNotification). Letting this code here decreases readability and also mixtures two different things (i.e. the actual handler and the notification push).

Sol3iman replied:
I created a helper function called sendPushNotification

0xZenus commented on src/mappings/factory/handleIrsInstanceDeployed.ts:
```diff
@@ -52,6 +55,36 @@ function handleIrsInstanceDeployed(event: IrsInstance): void {
+  const secret = 'null';
+  const cta = 'https://app.voltz.xyz/';
+
+  const notification = `{
```
Either that, or I would move this to a function in the same file (i.e. pushIRSDeployedNotification). Letting this code here decreases readability and also mixtures two different things (i.e. the actual handler and the notification push).

Sol3iman replied:
I created a helper function called sendPushNotification

Sol3iman committed:
feat: move sending notification to helper function in utilities
	- set image, secret, and cta fields to default values

Sol3iman committed:
feat: create a config variable NOTIFICATION_CONFIGS
	- tidy up the helper function sendPushNotification

Sol3iman commented on src/mappings/notifications/EPNSNotification.ts:
```diff
@@ -0,0 +1,31 @@
+import { BigInt, log } from '@graphprotocol/graph-ts';
```
moved it to /mappings/notifications

Sol3iman commented on src/mappings/notifications/EPNSNotification.ts:
```diff
@@ -0,0 +1,31 @@
+  }
+  epnsNotificationCounter.totalCount = epnsNotificationCounter.totalCount.plus(BigInt.fromI32(1));
+
+  const count = epnsNotificationCounter.totalCount.toHexString();
```
This is all from a boilerplate template

Sol3iman committed:
feat: added missing epnsNotification exporter into its handler function
	- renamed EPNSNotifications to handleEPNSNotifications
	- added missing exporter
	- added missing index file in /notifications
	- replaced the goerli subgraphID with the dedicated test subgraph

Sol3iman committed:
fix: write correct path in handleEPNSNotifications for schema.ts

Sol3iman committed:
fix: refactored files to handle AssemblyScript compiler
	- getProtocolPrefix was refactored from switch case to if statements
	- NOTIFICATION_CONFIGS was refactored into nested list
	- sendPushNotification utility was refactored to remove default values

Sol3iman committed:
fix: goerli channel address was missing 2 at the end

Sol3iman committed:
feat: refactored folder structure and added eip155:chainId to recipient field
	- EPNSNotification restored and put in utilities
	- sendEPNSNotification function name restored

0xZenus commented on src/mappings/factory/handleIrsInstanceDeployed.ts:
```diff
@@ -52,6 +55,22 @@ function handleIrsInstanceDeployed(event: IrsInstance): void {
   amm.termEndTimestamp = event.params.termEndTimestampWad;
   amm.tickSpacing = BigInt.fromI32(event.params.tickSpacing);
   amm.save();
+
```
Could you place the code below into a separate function (that stays below this handler in the file)?
So we separate the handler logic from the notification logic.

Sol3iman committed:
feat: this version of the push alerts code successfully sends notifications to the recipient

Sol3iman committed:
feat: working subgraph deployment showing dateTime formatted dates in deployment notifications
	- Added unix to date conversion utility

Sol3iman committed:
feat: working subgraph deployment for notifs
	- getProtocolPrefix explicit export default
	- index.ts now contains getProtocolPrefix and convertUnixToDate
	- import paths for getProtocolPrefix and convertUnixToDate in handleIrsInstanceDeployed and handleLiquidatedPosition updated to reflect updated index.ts file

Sol3iman committed:
feat: add convertMarginUnits utility to convert margin amount to correct decimals in notif message
	- add new constant MIL_BI to represent 1e6
	- add convertMarginUnits to index.ts
	- use convertMarginUnits in handleLiquidatePosition to convert margin.amount
	- subgraph deployment up to this commit allows successful notif sending
